pkg/syncer/spec/dns/dns_process.go

```diff
@@ -239,3 +251,48 @@ func hasAtLeastOneReadyAddress(endpoints *corev1.Endpoints) bool {
 	}
 	return false
 }
+
+func (d *DNSProcessor) mayUpdate(ctx context.Context, name string) (bool, error) {
+	logger := klog.FromContext(ctx)
+
+	deployment, err := d.deploymentLister.Deployments(d.dnsNamespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		return false, nil
+	}
+
+	needsUpdate := false
+	c := findContainer(deployment, "kcp-dns")
+
+	if c == nil {
+		// corrupted deployment. Trying to recover
+		expected := MakeDeployment(name, d.dnsNamespace, d.dnsImage)
+		deployment.Spec = expected.Spec
+		needsUpdate = true
+	} else if c.Image != d.dnsImage {
+		c.Image = d.dnsImage
```

Comment:
because of the minimalist approach and also because I didn't want to replace the entire Spec and potentially removed sidecars (among other things). 

In general I feel it's best to only update what fields the syncer controller owns, which is more than `dnsImage`, but less than `spec`. The idea is to cover more fields as needed.

Reply:
OK